pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -4278,16 +4298,33 @@ func mapNodeNameToPrivateDNSName(nodeName types.NodeName) string {
 }
 
 // mapInstanceToNodeName maps a EC2 instance to a k8s NodeName, by extracting the PrivateDNSName
-func mapInstanceToNodeName(i *ec2.Instance) types.NodeName {
-	return types.NodeName(aws.StringValue(i.PrivateDnsName))
+func mapInstanceToNodeName(i *ec2.Instance, explicitNames bool) (types.NodeName, error) {
+	if explicitNames {
+		for _, tag := range i.Tags {
+			key := aws.StringValue(tag.Key)
+			if key == TagNameKubernetesNodeNamePrefix {
+				value := aws.StringValue(tag.Value)
+				if len(value) > 0 {
+					return types.NodeName(aws.StringValue(tag.Value)), nil
+				}
+				return types.NodeName(""), fmt.Errorf("tag %s was found for instance %s, but had no value", TagNameKubernetesNodeNamePrefix, aws.StringValue(i.InstanceId))
+			}
+		}
+		return types.NodeName(""), fmt.Errorf("tag %s not found for instance %s", TagNameKubernetesNodeNamePrefix, aws.StringValue(i.InstanceId))
+	}
+	return types.NodeName(aws.StringValue(i.PrivateDnsName)), nil
 }
 
 // Returns the instance with the specified node name
 // Returns nil if it does not exist
 func (c *Cloud) findInstanceByNodeName(nodeName types.NodeName) (*ec2.Instance, error) {
 	privateDNSName := mapNodeNameToPrivateDNSName(nodeName)
+	nodeFilter := "private-dns-name"
+	if c.cfg.Global.ExplicitNodeNames {
+		nodeFilter = fmt.Sprintf("tag:%s", TagNameKubernetesNodeNamePrefix)
+	}
 	filters := []*ec2.Filter{
-		newEc2Filter("private-dns-name", privateDNSName),
+		newEc2Filter(nodeFilter, privateDNSName),
```

Comment:
Here we would have to attempt with the tag filter, check to see if any instances were returned, then try private DNS.